core/iwasm/libraries/libc-wasi/libc_wasi.cmake

```diff
@@ -11,3 +11,23 @@ include_directories(${LIBC_WASI_DIR}/sandboxed-system-primitives/include
 file (GLOB_RECURSE source_all ${LIBC_WASI_DIR}/*.c )
 
 set (LIBC_WASI_SOURCE ${source_all})
+
+file(WRITE "${PROJECT_BINARY_DIR}/try_compile_clock_nanosleep.c"
+"   #include <time.h>
+    void mysleep_ms(int millisec)
+    {
+        struct timespec res;
+        res.tv_sec = millisec/1000;
+        res.tv_nsec = (millisec*1000000) % 1000000000;
+        clock_nanosleep(CLOCK_MONOTONIC, 0, &res, NULL);
+    }
+    int main () { return 0; }
+")
+
+try_compile(WAMR_HAS_CLOCK_NANOSLEEP
```

Comment:
Hi, it seems that cmake already has `check_function_exists` function to do the same work, no need to use `try_compile` (which generates files under build folder and is not so friendly): https://cmake.org/cmake/help/v3.14/module/CheckFunctionExists.html

And by using `check_function_exists`, we only need to modify `libc_wasi.cmake` and `sandboxed-system-primitives/src/ssp_config.h`, no need to modify other two files. It works on building iwasm for linux and linux-sgx platforms. Here is the patch I tested, could you test whether it works for you:
```diff
diff --git a/core/iwasm/libraries/libc-wasi/libc_wasi.cmake b/core/iwasm/libraries/libc-wasi/libc_wasi.cmake
index d72c42a0..8549ae98 100644
--- a/core/iwasm/libraries/libc-wasi/libc_wasi.cmake
+++ b/core/iwasm/libraries/libc-wasi/libc_wasi.cmake
@@ -11,3 +11,10 @@ include_directories(${LIBC_WASI_DIR}/sandboxed-system-primitives/include
 file (GLOB_RECURSE source_all ${LIBC_WASI_DIR}/*.c )

 set (LIBC_WASI_SOURCE ${source_all})
+
+include (CheckFunctionExists)
+check_function_exists(clock_nanosleep WAMR_HAS_CLOCK_NANOSLEEP)
+
+if (WAMR_HAS_CLOCK_NANOSLEEP)
+  add_definitions(-DCLOCK_NANOSLEEP_COMPILES)
+endif()
diff --git a/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/ssp_config.h b/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/ssp_config.h
index d5babd02..094d52e4 100644
--- a/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/ssp_config.h
+++ b/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/ssp_config.h
@@ -41,7 +41,7 @@
 #endif

 #if !defined(__APPLE__) && !defined(__FreeBSD__) && !defined(__EMSCRIPTEN__) \
-    && !defined(ESP_PLATFORM)
+    && !defined(ESP_PLATFORM) && defined(CLOCK_NANOSLEEP_COMPILES)
 #define CONFIG_HAS_CLOCK_NANOSLEEP 1
 #else
 #define CONFIG_HAS_CLOCK_NANOSLEEP 0
```

Reply:
Seems better, I checked on linux and it seems fine. I'll check it out on a device without clock_nanosleep tommorrow in the office

Reply:
I checked and it worked this way. BUT I found that I was wrong in the first place, because clock_nanosleep was available on the platform but it used netbsd https://man.netbsd.org/clock_nanosleep.2 and even function was defined it returned ENOSYS

So it means that try_compile and checkFunctionExists detected absense of the function wrong, so I decided to go with a flag because its simplier and more explicit solution even it's less elegant

Reply:
But now I don't see any logs in CI failing jobs :thinking:

Reply:
> I checked and it worked this way. BUT I found that I was wrong in the first place, because clock_nanosleep was available on the platform but it used netbsd https://man.netbsd.org/clock_nanosleep.2 and even function was defined it returned ENOSYS

while i have no idea about Roku 4200X,
i'm sure NetBSD has had clock_nanosleep for a decade.

Reply:
> But now I don't see any logs in CI failing jobs 🤔

github is broken today.
https://www.githubstatus.com/

Reply:
> i'm sure NetBSD has had clock_nanosleep for a decade.
And it does have it, it compiles. And that's why idea of using `try_compile` actually wasn't solving the problem I met.
clock_nanosleep on this platform returns ENOSYS, which is correct behavior according to the NetBSD doc, but it means that it doesn't work.

Now I just switched to a simplier solution

Reply:
> > i'm sure NetBSD has had clock_nanosleep for a decade.
> 
> And it does have it, it compiles. And that's why idea of using `try_compile` actually wasn't solving the problem I met. `clock_nanosleep` on this platform returns ENOSYS, which is correct behavior according to the NetBSD doc, but it means that implementation on this platform is stub and doesn't do anything.

i meant NetBSD has had working clock_nanosleep.
i guess it's just listing theoretical values.
it isn't even NetBSD-specific. cf. https://pubs.opengroup.org/onlinepubs/007908799/xsh/nanosleep.html

> 
> Now I just switched to a simplier solution

ok.